### PR TITLE
[media] Create IamfMimeUtil

### PR DESCRIPTION
**Pull request by osagie98**
b/356704307

### CODE REVIEW

File: starboard/shared/starboard/media/codec_util.cc [hunk 1]
```diff
@@ -107,13 +108,102 @@ SbMediaAudioCodec GetAudioCodecFromString(const char* codec,
...
+bool IsIamfMimeType(std::string mime_type) {
+  // Reference: Immersive Audio Model and Formats;
+  //            v1.0.0
+  //            6.3. Codecs Parameter String
+  // (https://aomediacodec.github.io/iamf/v1.0.0-errata.html#codecsparameter)
+  if (mime_type.find("iamf") == std::string::npos) {
+    return false;
+  }
+
+  constexpr int kMaxIamfCodecIdLength =
```
Comment (borongc) [inline starboard/shared/starboard/media/codec_util.cc:128]: Although addition is a cheap operation, should we consider directly put the total number here, and then comment how it is calculated?

File: starboard/common/string.h [hunk 2]
```diff
@@ -103,6 +104,23 @@ static SB_C_FORCE_INLINE int strlcat(CHAR* dst, const CHAR* src, int dst_size) {
...
+inline std::vector<std::string> SplitString(std::string& input,
+                                            char delimiter) {
+  std::vector<std::string> output;
+  if (input.find(delimiter) == std::string::npos) {
+    return output;
+  }
+
+  std::stringstream stream(input);
+  std::string token;
+  while (std::getline(stream, token, delimiter)) {
```
Comment (borongc) [inline starboard/common/string.h:118]: Just curious if this is the most efficient way to split strings.

File: starboard/common/string_test.cc [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#include <string>
+#include <vector>
+
+#include "starboard/common/string.h"
+
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace starboard {
+namespace {
+
```
Comment (borongc) [inline starboard/common/string_test.cc:24]: Consider to add more test cases, such as `EmptyString`. Refer to https://source.chromium.org/chromium/chromium/src/+/main:base/strings/string_split_unittest.cc.

As we will use this `SplitString()` to parse our MIME string, we should consider example MIME strings (especially IAMF), e.g., `iamf.000.000.mp4a.40.2`.

File: starboard/shared/starboard/media/codec_util.cc [hunk 4]
```diff
@@ -107,13 +108,101 @@ SbMediaAudioCodec GetAudioCodecFromString(const char* codec,
...
+  // +1  delimiting period.
+  // +9  The remaining string is one of "Opus", "mp4a.40.2", "fLaC", or "ipcm".
+  constexpr int kMaxIamfCodecIdLength = 22;
+
+  if (mime_type.size() > kMaxIamfCodecIdLength) {
+    return false;
+  }
+
+  std::vector<std::string> vec = SplitString(mime_type, '.');
+  if (vec.size() != 4 && vec.size() != 6) {
```
Comment (borongc) [inline starboard/shared/starboard/media/codec_util.cc:142]: Maybe we should specify why `4` and `6` here as a comment.

File: starboard/shared/starboard/media/codec_util.cc [hunk 5]
```diff
@@ -107,13 +108,101 @@ SbMediaAudioCodec GetAudioCodecFromString(const char* codec,
...
+  }
+
+  if (vec[0] != "iamf") {
+    return false;
+  }
+
+  // The primary profile string should be three digits, and should be between 0
+  // and 255 inclusive.
+  int primary_profile;
+  std::stringstream stream(vec[1]);
```
Comment (borongc) [inline starboard/shared/starboard/media/codec_util.cc:153]: vec[1] is std::string, should we consider to use `std::stoi()` to convert string to integer here?

File: starboard/shared/starboard/media/codec_util.cc [hunk 6]
```diff
@@ -107,13 +108,101 @@ SbMediaAudioCodec GetAudioCodecFromString(const char* codec,
...
+  // 0 and 255 inclusive.
+  stream = std::stringstream(vec[2]);
+  int additional_profile;
+  stream >> additional_profile;
+  if (stream.fail() || vec[2].size() != 3 || additional_profile > 255) {
+    return false;
+  }
+
+  // The codec string should be one of "Opus", "mp4a", "fLaC", or "ipcm".
+  std::string codec = vec[3];
```
Comment (borongc) [inline starboard/shared/starboard/media/codec_util.cc:169]: Maybe consider to refactor the following code to switch-case, so it is more readable.

For example:
```
switch(codec) {
  case "Opus":
  case "fLaC":
  case "ipcm":
    return true;
  default:
    // handle "mp4a.40.2" and check its format here
    // return true if it meets
}

return false;
```

File: starboard/shared/starboard/media/codec_util_test.cc [hunk 7]
```diff
@@ -263,6 +263,48 @@ TEST(CodecUtilTest, DoesNotParse1AsPcmForNonWavSubtypes) {
...
+            kSbMediaAudioCodecIamf);
+  EXPECT_EQ(GetAudioCodecFromString("iamf.000.000.mp4a.40.2", ""),
+            kSbMediaAudioCodecIamf);
+  EXPECT_EQ(GetAudioCodecFromString("iamf.000.000.fLaC", ""),
+            kSbMediaAudioCodecIamf);
+  EXPECT_EQ(GetAudioCodecFromString("iamf.000.000.ipcm", ""),
+            kSbMediaAudioCodecIamf);
+}
+
+TEST(CodecUtilTest, IsIamfMimeType) {
```
Comment (borongc) [inline starboard/shared/starboard/media/codec_util_test.cc:281]: Consider to add cases where the length of resulting string vector is not 4, like `iamf.xxx.Opus`, `iamf.1xx.2yy`.

File: starboard/common/string_test.cc (hunk 3 above)
Comment (osagie98) [reply starboard/common/string_test.cc:24]: Done

File: starboard/shared/starboard/media/codec_util.cc (hunk 4 above)
Comment (osagie98) [reply starboard/shared/starboard/media/codec_util.cc:141]: Added the comment. Seems that adding a special case is pretty specific to this use case, I'm not sure if it'd see any use outside of this function to justify adding it in. Happy to discuss it.

File: starboard/shared/starboard/media/codec_util.cc (hunk 6 above)
Comment (osagie98) [reply starboard/shared/starboard/media/codec_util.cc:168]: Unfortunately switch statements can't work with strings, only values resolving to int.

File: starboard/shared/starboard/media/codec_util_test.cc (hunk 7 above)
Comment (osagie98) [reply starboard/shared/starboard/media/codec_util_test.cc:278]: Done

File: starboard/shared/starboard/media/codec_util.cc [hunk 8]
```diff
@@ -106,12 +107,100 @@ SbMediaAudioCodec GetAudioCodecFromString(const char* codec,
   if (is_wav && strcmp(codec, "1") == 0) {
     return kSbMediaAudioCodecPcm;
   }
-  if (strcmp(codec, "iamf") == 0 || strncmp(codec, "iamf.", 5) == 0) {
+  if (strcmp(codec, "iamf") == 0 || IsIamfMimeType(codec)) {
```
Comment (jasonzhangxx) [inline starboard/shared/starboard/media/codec_util.cc:110]: We don't check the validity of the codec string in this function.

The best way to check the validity of the iamf codec string I can think of is to check it in SbMediaIsAudioSupported(). We can add an iamf_util, and check the validity of codec string while deciding if the sub codec can be supported.

File: starboard/common/string.h [hunk 9]
```diff
@@ -101,6 +101,37 @@ static SB_C_FORCE_INLINE int strlcat(CHAR* dst, const CHAR* src, int dst_size) {
          dst_length;
 }
 
+// Splits a string on a char delimiter. Returns an empty vector if the delimiter
+// is not found
```
Comment (xiaomings) [inline starboard/common/string.h:105]: Consider double checking if returning an empty vector when the delimiter is not found is a common practice.

File: starboard/common/string.h [hunk 10]
```diff
@@ -101,6 +101,37 @@ static SB_C_FORCE_INLINE int strlcat(CHAR* dst, const CHAR* src, int dst_size) {
...
+inline std::vector<std::string> SplitString(std::string& input,
```
Comment (xiaomings) [inline starboard/common/string.h:106]: const std::string&

File: starboard/shared/starboard/media/codec_util_test.cc [hunk 11]
```diff
@@ -263,6 +263,18 @@ TEST(CodecUtilTest, DoesNotParse1AsPcmForNonWavSubtypes) {
   EXPECT_EQ(GetAudioCodecFromString("1", "webm"), kSbMediaAudioCodecNone);
 }
 
+TEST(CodecUtilTest, ParsesIamfCodec) {
```
Comment (xiaomings) [inline starboard/shared/starboard/media/codec_util_test.cc:266]: Just to double check if this makes use of IamfMimeUtil?

File: starboard/shared/starboard/media/iamf_util.h [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+namespace shared {
+namespace starboard {
+namespace media {
+
+enum IamfSubstreamCodec {
+  kIamfSubstreamCodecOpus,
+  kIamfSubstreamCodecMp4a,
+  kIamfSubstreamCodecFlac,
+  kIamfSubstreamCodecIpcm,
+  kIamfSubstreamCodecUnknown
```
Comment (xiaomings) [inline starboard/shared/starboard/media/iamf_util.h:31]: Unless this is a map of spec defined somewhere, consider moving kIamfSubstreamCodecUnknown to the first, so when the value is zero initialized, it's initialized to kIamfSubstreamCodecUnknown.

File: starboard/shared/starboard/media/iamf_util.h [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+// Always check is_valid() before calling the getter functions.
+class IamfMimeUtil {
+ public:
+  explicit IamfMimeUtil(const std::string& mime_type);
+
+  bool is_valid() const {
+    return profile_ <= kIamfProfileMax &&
+           substream_codec_ != kIamfSubstreamCodecUnknown;
+  }
+  int profile() const { return profile_; }
```
Comment (xiaomings) [inline starboard/shared/starboard/media/iamf_util.h:51]: Add DCHECK(is_valid()), and the same to below.

File: starboard/shared/starboard/media/iamf_util.cc [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+namespace media {
+
+IamfMimeUtil::IamfMimeUtil(const std::string& mime_type) {
+  // Reference: Immersive Audio Model and Formats;
+  //            v1.0.0
+  //            6.3. Codecs Parameter String
+  // (https://aomediacodec.github.io/iamf/v1.0.0-errata.html#codecsparameter)
+  if (mime_type.find("iamf") == std::string::npos) {
+    return;
+  }
```
Comment (xiaomings) [inline starboard/shared/starboard/media/iamf_util.cc:35]: This is probably redundant to the check in line 57?  And if we do want to check, we should check that the return value is 0.

File: starboard/shared/starboard/media/iamf_util.cc [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+  stream = std::stringstream(vec[2]);
+  int additional_profile;
+  stream >> additional_profile;
+  if (stream.fail() || vec[2].size() != 3 || additional_profile > 255) {
+    return;
+  }
+
+  // The codec string should be one of "Opus", "mp4a", "fLaC", or "ipcm".
+  std::string codec = vec[3];
+  if (codec.size() != 4 || ((codec != "Opus") && (codec != "mp4a") &&
```
Comment (xiaomings) [inline starboard/shared/starboard/media/iamf_util.cc:81]: The check `codec.size() != 4` is redundant.

File: starboard/shared/starboard/media/iamf_util.cc [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    if (stream.fail() || vec[4].size() != 2 || object_type_indication != 40) {
+      return;
+    }
+
+    stream = std::stringstream(vec[5]);
+    int audio_object_type;
+    stream >> audio_object_type;
+    if (stream.fail() || vec[5].size() != 1 || audio_object_type != 2) {
+      return;
+    }
```
Comment (xiaomings) [inline starboard/shared/starboard/media/iamf_util.cc:106]: Shall we just check `vec[4] == '40' && vec[5] == '2'`?

**Comment by osagie98** [comment]
This PR will be close. The 25.lts PR is here: https://github.com/youtube/cobalt/pull/4244

**Pull request closed without merging**